### PR TITLE
Issue #6489: correct one-char loop variable in LocalVariableNameCheck

### PR DESCRIPTION
**Pull request by strkkk**
Issue: #6489 
Case for one-char loop variable is treated differently now, test case was updated.

Regression report will be soon.

### CODE REVIEW

**Comment by rnveach** [comment]
Travis is having some issues right now and can be ignored.
Appveyor looks valid and is reporting test failures.

wercker is showing regression failures in XWiki:
````
[INFO] --- maven-checkstyle-plugin:3.0.0:check (default) @ xwiki-commons-tool-test-simple ---
[INFO] Starting audit...
[ERROR] /pipeline/source/.ci-temp/xwiki-commons/xwiki-commons-tools/xwiki-commons-tool-test/xwiki-commons-tool-test-simple/src/main/java/org/xwiki/test/AllLogRule.java:202:18: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalVariableName]
[ERROR] /pipeline/source/.ci-temp/xwiki-commons/xwiki-commons-tools/xwiki-commons-tool-test/xwiki-commons-tool-test-simple/src/main/java/org/xwiki/test/AllLogRule.java:241:22: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalVariableName]
[ERROR] /pipeline/source/.ci-temp/xwiki-commons/xwiki-commons-tools/xwiki-commons-tool-test/xwiki-commons-tool-test-simple/src/main/java/org/xwiki/test/junit5/LogCaptureExtension.java:149:18: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalVariableName]
[ERROR] /pipeline/source/.ci-temp/xwiki-commons/xwiki-commons-tools/xwiki-commons-tool-test/xwiki-commons-tool-test-simple/src/main/java/org/xwiki/test/junit5/LogCaptureExtension.java:186:22: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalVariableName]
[INFO] There are 4 errors reported by Checkstyle 8.20-SNAPSHOT with checkstyle.xml ruleset.
[ERROR] src/main/java/org/xwiki/test/AllLogRule.java:[202,18] (naming) LocalVariableName: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'.
[ERROR] src/main/java/org/xwiki/test/AllLogRule.java:[241,22] (naming) LocalVariableName: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'.
[ERROR] src/main/java/org/xwiki/test/junit5/LogCaptureExtension.java:[149,18] (naming) LocalVariableName: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'.
[ERROR] src/main/java/org/xwiki/test/junit5/LogCaptureExtension.java:[186,22] (naming) LocalVariableName: Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'.
Audit done.
````

Please review xwiki violations. If they are valid, a PR will need to be given to xwiki project with a fix for the failures. Once PR is accepted we can re-run wercker here and it should pass for xwiki and continue on.

**Comment by strkkk** [comment]
@rnveach They use default settings, so these violations are valid because these variable names are not allowed by default.

Other thing is that message "Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'." is a bit confusing because this variable name actually **matches** given pattern (it was root cause of the bug).
Options are, in my opinion:
1. Ignore it
2. Add new violation pattern, e.g. "One char loop variable names are not allowed"

**Comment by romani** [comment]
> Other thing is that message "Name 'i' must match pattern '^[a-z][a-zA-Z0-9]*$'." is a bit confusing because this variable name actually matches given pattern (it was root cause of the bug).

good catch ! we need to fix message.

> Add new violation pattern, e.g. "One char loop variable names are not allowed".

I like this option. @rnveach , do you agree on wording ?
It is not a problem for Check to have multiple messages.
Please introduce new message and use google translator to translate to other languages (just follow build violations, it will guide you).

----

It is not very pleasant, but we need to resolve issues in xwiki and simple-binary-encoding, by providing them PR. It will be first good user experience of fix (before release!!!). Please send them PRs to fix problems, ones they merge them we can merge your PR.

**Comment by strkkk** [comment]
@romani original [issue](https://github.com/checkstyle/checkstyle/issues/6489) requires clarification, please check it.
I will update my pr accordingly.

**Comment by romani** [comment]
@strkkk , please review reply https://github.com/checkstyle/checkstyle/issues/6489#issuecomment-487300833

**Comment by rnveach** [comment]
@romani Please review first to check requirements.

**Comment by strkkk** [comment]
@romani Please take a look at this PR

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java [hunk 1]
```diff
@@ -96,10 +96,11 @@
  * <p>
  * An example of how to configure the check to allow one character variable name in
  * <a href="https://docs.oracle.com/javase/tutorial/java/nutsandbolts/for.html">
- * initialization expressions</a> in FOR loop:
+ * initialization expressions</a> in FOR loop, where regexp allows 2 or more chars:
  * </p>
  * <pre>
  * &lt;module name="LocalVariableName"&gt;
+ *   &lt;property name="format" value="^[a-z][_a-zA-Z0-9]+$"/&gt;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: please do us a favor to update documentation a bit more.
Please make it like:
`An example of how to configure the check ...`
following with
```
Code Example:
class MyClass {
  void MyMethod() {
    int goodName = 0;
    int i = 1; // violation
    for (int var = 1; var < 10; var++) { //OK
      int j = 1; // violation
    }
  }
}
```

xdoc should also be update, it should match javadoc.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localvariablename/InputLocalVariableNameOneCharInitVarName.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
 {
 	public void fooMethod()
 	{
-		for(int i = 1; i <10; i++) {
+		for(int i = 1, J = 1; i <10; i++) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localvariablename/InputLocalVariableNameOneCharInitVarName.java]: update in tests is very small, I expected more updates.

Please provide links to lines in Input files which covers all cases that we discussed in issue and you have in doc examples.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java (hunk 1 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: Current example shows case when regex allows 2+ chars, and property is used to make an exception.

Did you mean to add one more example with allowOneCharVarInForLoop = false and regex for 3+ chars or adjust existing one?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localvariablename/InputLocalVariableNameOneCharInitVarName.java (hunk 2 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/localvariablename/InputLocalVariableNameOneCharInitVarName.java]: I added test cases for capital letters in for init statement in for-each and usual for.
Documentation was updated with similar examples

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java (hunk 1 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: I added one more example. Please take a look.

**Comment by strkkk** [comment]
Appveyor [pipeline](https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/24524358/job/ct4ph7bf9hldy9si) is failed because of some infra issue
>[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.7.1:site (default-site) on project checkstyle: Error parsing 'C:\projects\checkstyle\src\xdocs\anttask.xml.vm': line [-1] Error validating the model: No such host is known (www.w3.org) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.7.1:site (default-site) on project checkstyle: Error parsing 'C:\projects\checkstyle\src\xdocs\anttask.xml.vm': line [-1] Error validating the model: No such host is known (www.w3.org)

@rnveach can you tell me what is wrong with wercker build? I have no access to build log.

**Comment by rnveach** [comment]
@strkkk wercker failure should be because you aren't on the lastest master as we added a new check recently.
> CheckstyleException: Unable to instantiate 'MissingJavadocMethod' class, it is also not possible to instantiate it as ... MissingJavadocMethod

I restarted appveyor.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java [hunk 3]
```diff
@@ -92,32 +102,87 @@
...
+     * for (int K = 1; K &lt; 10; K++) { // OK
+     *     int j = 1; // violation
+     * }
+     * List list = new ArrayList();
+     * for (Object o : list) { // OK
+     *     int j = 1; // violation
+     * }
+     * for (Object O : list) { // OK
+     *     int j = 1; // violation
+     * }
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: @strkkk , is it something(CI) force you to update doc on field ?
just curious ...

I do not think we should have very extensive jaavadocs over properties, we have dedicated area in Class javadoc for this, no need to copy examples in other parts.

@pbludov , fyi.

**Comment by romani** [comment]
@rnveach , please finalize

**Comment by romani** [comment]
@strkkk , 

> Regression report will be soon.

please do.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: I agree with @romani, this long example shouldn't be in property description. That is why it is being forced on field through our xdoc / javadoc alignment process.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: @romani yes, there is fail if javadoc is updated only for class, not for field/setter

**Comment by strkkk** [comment]
@romani @rnveach 
report is [here](https://strkkk.github.io/6489/index.html)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java]: if any code changes be required, I would be good to remove code example from property description. Examples section should be enough, and we have link to java documentation , it is enough.

**Comment by romani** [comment]
report is good.
@rnveach , please finalize.

**Comment by strkkk** [comment]
@rnveach please take a look

**Comment by romani** [comment]
@rnveach , please take a look.

**Pull request merged**